architecture/2024-08-resource-types-registration.md

```diff
@@ -0,0 +1,580 @@
+
+The design proposal includes axes for configuring location affinity:
+
+- `System.Resources/resourceProviders` includes a list of supported locations with an optional `address` in each location for the resource provider.
+- `System.Resources/resourceProviders/resourceTypes` includes a list of locations where the resource type is supported and the API versions supported in each of those locations.
+
+**Open Question:** There's a few possible designs for this that might make sense. Right now location shows up as a concept in both `resourceProviders` and `resourceTypes`. It would be possible to move all of this to `resourceProviders`. It would also be possible to build another resource type `System.Resources/resourceProviders/locations` for the location configuration. Let's discuss.
+
+### Architecture Diagram
+
+Resource type registration and querying is implemented in UCP. Any client may call the registration APIs or query APIs including resource providers.
+
+<center>
+
+![architecture diagram](./2024-08-user-defined-types-registration/registration.png)
+
+**Fig: Diagram of resource type registration and querying** [link](https://excalidraw.com/#json=FzQQaxv1d-vlmmDBAeOfc,UUXIasjO8p27hN0AB8qeiw)
+
+</center>
+
+### Detailed Design
+
+Since this is our first nested resource type, a basic primer on nested types is needed. A nested resource type exists *inside* another resource type. A resource type is that is not nested is *top-level*. These relationships are also sometimes referred to as a parent and child. 
+
+The following statements are all accurate:
+
+- The `System.Resources/resourceProviders` type is the parent of `System.Resources/resourceProviders/resourceTypes`. 
+- The `System.Resources/resourceProviders/resourceTypes` type is the child of `System.Resources/resourceProviders`.
+- The `System.Resources/resourceProviders/resourceTypes` type is nested within `System.Resources/resourceProviders`.
+- The `System.Resources/resourceProviders` type is a top-level resource type.
+
+Any resource type with two segments (separated by `/`) is a top-level resource type. Any resource type with 3 or more segments (separate by `/`) is a nested resource type.
+
+The resource ids (and URLs) are written like the following:
+
+- `/planes/radius/{plane}/providers/System.Resources/resourceProviders` - List all resource providers.
+- `/planes/radius/{plane}/providers/System.Resources/resourceProviders/Contoso.Example` - Get the Contoso.Example resource provider.
+- `/planes/radius/{plane}/providers/System.Resources/resourceProviders/Contoso.Example/resourceTypes` - List all resource types inside the `Contoso.Example` resource provider/namespace.
+- `/planes/radius/{plane}/providers/System.Resources/resourceProviders/Contoso.Example/resourceTypes/contosoBuses` - Gets the `contosoBuses` resource type inside the `Contoso.Example` resource provider/namespace.
```

Comment:
I think you have explained it above, but I'm struggling a bit to understand the need for `System.Resources/resourceProviders/<name>/resourceTypes` in the overall id. Would be great to discuss this..

Reply:
Discussed in the meeting. My action item is to provide more examples. 

Answer to this question:

`/planes/radius/{plane}/providers/System.Resources/resourceProviders/Contoso.Example/resourceTypes/contosoBuses` represents the **type definition** for `Contoso.Example/contosoBuses`.

The resource IDs and paths of `Contoso.Example/contosoBuses` will look like our other resources: eg. `/planes/radius/{plane}/resourceGroups/foo/providers/Contoso.Example/contosoBuses/my-resource`